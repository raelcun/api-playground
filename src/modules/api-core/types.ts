import * as t from 'io-ts'
import Koa from 'koa'

import { RolesV } from '@lib/rbac'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
export type KoaContext2<T> = Koa.ParameterizedContext<Koa.DefaultState, { request: { body: T } }>

type a = Koa.Middleware<Koa.DefaultState, { request: { body: T } }>

export type Middleware<T> = (ctx: KoaContext<T>, next: () => Promise<void>) => Promise<void>

export const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>
