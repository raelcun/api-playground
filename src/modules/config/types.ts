import * as t from 'io-ts'

import { LoggingConfig } from '@lib/logger'
import { DeepPartial, NonEmptyString } from '@lib/utils'

export interface FullConfig {
  isProduction: boolean
  env: string
  server: {
    jwtSecret: string
    securePort: number
    insecurePort: number
    enableSSL: boolean
  }
  logging: LoggingConfig
  aws: {
    region: string
    credentials: {
      accessKeyId: string
      secretAccessKey: string
    }
  }
}

export type PartialConfig<T> = DeepPartial<T>

export type ConfigProvider = () => FullConfig

export const envVarsV = t.type({
  NODE_ENV: NonEmptyString,
  JWT_SECRET: NonEmptyString,
  AWS_ACCESS_KEY_ID: NonEmptyString,
  AWS_SECRET_ACCESS_KEY: NonEmptyString,
})
export type EnvVars = t.TypeOf<typeof envVarsV>
