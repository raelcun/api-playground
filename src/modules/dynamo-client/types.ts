import { DocumentClient } from 'aws-sdk/clients/dynamodb'
import { AWSError } from 'aws-sdk/lib/error'
import { Response } from 'aws-sdk/lib/response'
import { taskEither as TE } from 'fp-ts'

import { Err } from '@modules/error/types'
import { ChangeTypeOfKeys } from '@root/utils/types'

export interface DynamoClient<T> {
  put: (
    params: ChangeTypeOfKeys<DocumentClient.PutItemInput, 'Item', T>,
  ) => TE.TaskEither<Err, Response<DocumentClient.PutItemOutput, AWSError>>
  delete: (
    params: DocumentClient.DeleteItemInput,
  ) => TE.TaskEither<Err, Response<DocumentClient.DeleteItemOutput, AWSError>>
}
