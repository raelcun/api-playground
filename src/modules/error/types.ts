export interface Err<Code extends string = string> {
  code: Code
  message?: string
}
