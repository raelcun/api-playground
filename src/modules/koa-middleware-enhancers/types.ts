import Koa from 'koa'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
