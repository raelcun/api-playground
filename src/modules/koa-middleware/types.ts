import { KoaContext } from '../koa-middleware-enhancers'

export type Middleware<T> = (
  ctx: KoaContext<T>,
  next: () => Promise<void>,
) => Promise<void>
