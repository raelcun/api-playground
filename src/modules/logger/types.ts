import { Logger as PinoLogger } from 'pino'

export type LogMethods = 'fatal' | 'trace' | 'error' | 'info' | 'debug' | 'warn'

export type Logger = Pick<PinoLogger, LogMethods | 'child'>

export type LoggerFactory = () => Logger
