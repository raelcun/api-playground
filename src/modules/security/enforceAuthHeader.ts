import { tryCatch2v, Either, left, right } from 'fp-ts/lib/Either'
import { verify } from 'jsonwebtoken'
import * as t from 'io-ts'
import { Task } from 'fp-ts/lib/Task'
import { TaskEither, fromEither } from 'fp-ts/lib/TaskEither'
import { IO } from 'fp-ts/lib/IO'
import { getConfig } from '../../config'
import { Actions, RolesV, Roles, Enforce } from '../rbac/types'
import { Error } from '../error/types'
import { Middleware } from '../../types'
import { getEnforcer } from '../rbac'
import { getSystemLogger } from '../logger'

const resolveAuthHeader = (headers: unknown): Either<Error, [string, string]> => {
  const getAuthHeader = (headers: unknown): Either<Error, string> =>
    t
      .type({ authorization: t.string })
      .decode(headers)
      .mapLeft(() => ({ code: 'INVALID_AUTH_HEADER', message: 'invalid auth header format' }))
      .map(({ authorization }) => authorization)

  const parseAuthHeaderParts = (parts: string[]): Either<Error, [string, string]> => {
    if (parts.length !== 2)
      return left({ code: 'INVALID_AUTH_HEADER', message: 'requires two parts' })
    if (parts[0] !== 'Bearer')
      return left({ code: 'INVALID_AUTH_HEADER', message: 'first part must be "Bearer"' })
    return right([parts[0], parts[1]])
  }

  return getAuthHeader(headers)
    .map(authorization => authorization.split(' '))
    .chain(parseAuthHeaderParts)
}

const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>

const verifyAndParseToken = (token: string): Either<Error, Token> => {
  return tryCatch2v(
    () => verify(token, getConfig().server.jwtSecret),
    () => ({ code: 'INVALID_TOKEN', message: 'failed to decode token' }),
  ).chain(tokenPayload =>
    tokenV
      .decode(tokenPayload)
      .mapLeft(() => ({ code: 'INVALID_TOKEN', message: 'token payload format invalid' })),
  )
}

export const enforceWithAuthHeaderInternal = (enforcerProvider: () => Promise<Enforce>) => <
  T extends keyof Actions,
  U extends Actions[T]
>(
  resource: T,
  actions: U[],
): Middleware<unknown> => async (ctx, next) => {
  const validateRoleWithEnforcer = (role: Roles): TaskEither<Error, undefined> =>
    new TaskEither(
      new Task(async () => {
        try {
          const enforcer = await enforcerProvider()
          const validationResult = await enforcer(role, resource, ...actions)
          return validationResult === false ? left({ code: 'UNAUTHORIZED' }) : right(undefined)
        } catch (e) {
          return left({ code: 'AUTHORIZATION_FAILED' })
        }
      }),
    )
  const onUnauthorized = () =>
    new IO(async () => {
      ctx.status = 401
      ctx.body = 'bad'
    })
  const onAuthorized = () =>
    new IO(async () => {
      await next()
    })

  const sideEffects = await fromEither(
    resolveAuthHeader(ctx.headers).chain(([, credentials]) => verifyAndParseToken(credentials)),
  )
    .chain(token => validateRoleWithEnforcer(token.role))
    .mapLeft(e => {
      getSystemLogger().trace(e)
      return e
    })
    .fold(onUnauthorized, onAuthorized)
    .run()
  await sideEffects.run()
}

export const enforceWithAuthHeader = enforceWithAuthHeaderInternal(getEnforcer)
