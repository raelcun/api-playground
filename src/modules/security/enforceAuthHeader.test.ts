import { createMockContext } from '@shopify/jest-koa-mocks'
import { sign } from 'jsonwebtoken'
import { Context } from 'koa'
import { Enforce, Roles } from '../rbac/types'
import { getConfig } from '../../config'
import { enforceWithAuthHeaderInternal, Token } from './enforceAuthHeader'

const createToken = (token: Token = { userId: 'foo', role: Roles.User }) =>
  sign(token, getConfig().server.jwtSecret)
const createDefaultMockContext = (token: Token = { userId: 'foo', role: Roles.User }) =>
  createMockContext({ headers: { authorization: `Bearer ${createToken(token)}` } })

const expectUnauthorized = (mockContext: Context, next: jest.Mock) => {
  expect(mockContext.status).toEqual(401)
  expect(mockContext.body).toMatchSnapshot()
  expect(next).not.toHaveBeenCalled()
}

describe('enforceWithAuthHeaderInternal', () => {
  test('should invoke next when authorized (via enforcer)', async () => {
    const mockEnforcer = () => Promise.resolve<Enforce>(async () => true)
    const mockContext = createDefaultMockContext()
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expect(next).toHaveBeenCalled()
  })

  test('should set status to 401 and set body when unauthorized (via enforcer)', async () => {
    const mockEnforcer = () => Promise.resolve<Enforce>(async () => false)
    const mockContext = createDefaultMockContext()
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if enforcer throws', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createDefaultMockContext()
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if no auth header', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext()
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if not using bearer token', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext({
      headers: { authorization: `NotBearer ${createToken()}` },
    })
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if not using bearer token', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext({
      headers: { authorization: `NotBearer ${createToken()}` },
    })
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if invalid auth header', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext({
      headers: { authorization: 'totallynotvalid' },
    })
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if invalid token', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext({
      headers: { authorization: 'Bearer totallynotvalid' },
    })
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })

  test('should be unauthorized if expired token', async () => {
    const mockEnforcer = () =>
      Promise.resolve<Enforce>(async () => {
        throw 'bad error'
      })
    const mockContext = createMockContext({
      headers: {
        authorization: `Bearer ${sign(
          { userId: 'foo', role: Roles.User },
          getConfig().server.jwtSecret,
        )}`,
      },
    })
    const next = jest.fn()
    await enforceWithAuthHeaderInternal(mockEnforcer)('account', ['viewAny'])(mockContext, next)
    expectUnauthorized(mockContext, next)
  })
})
