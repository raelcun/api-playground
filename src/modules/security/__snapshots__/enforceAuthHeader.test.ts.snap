// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enforceWithAuthHeaderInternal should be unauthorized if enforcer throws 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if expired token 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if invalid auth header 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if invalid token 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if invalid token payload 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if no auth header 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should be unauthorized if not using bearer token 1`] = `"bad"`;

exports[`enforceWithAuthHeaderInternal should set status to 401 and set body when unauthorized (via enforcer) 1`] = `"bad"`;
